### PR TITLE
feat(dev-infra): write outputs of command runs to ng-dev log file

### PR DESCRIPTION
**Pull request by josephperrott**
Creates infrastructure to write outputs of command runs to ng-dev log file.
Additionally, on commands which fail with an exit code greater than 1, an
error log file is created with the posix timestamp of the commands run time
as an identifier.

### CODE REVIEW

File: dev-infra/utils/console.ts [hunk 1]
```diff
@@ -155,3 +160,41 @@ function getLogLevel() {
...
+    // file for the command run failure.
+    if (code > 1) {
+      writeFileSync(join(getRepoBaseDir(), `.ng-dev.err-${now.getTime()}.log`), LOGGED_TEXT);
+    }
+  });
+}
+
+/** Write the provided text to the log file, prepending each line with the log level.  */
+function printToLogFile(logLevel: LOG_LEVELS, ...text: string[]) {
+  const logLevelText = `${LOG_LEVELS[logLevel]}:`.padEnd(7);
```
Comment (devversion) [inline dev-infra/utils/console.ts:198]: Optional nit: I'd propose extracting the amount of spacing into a constant for this file (that will increase readability)

File: dev-infra/utils/console.ts [hunk 2]
```diff
@@ -155,3 +160,41 @@ function getLogLevel() {
...
+  process.on('exit', (code: number) => {
+    /** Path to the log file location. */
+    const logFilePath = join(getRepoBaseDir(), '.ng-dev.log');
+    /** The current contents of the log file. */
+    const logContents = existsSync(logFilePath) ? readFileSync(logFilePath).toString() : '';
+    /**
+     * The new contents of the log file, the last 10,000 lines of the combined current and new
+     * logged lines.
+     */
+    const combinedLog = (logContents + LOGGED_TEXT).split('\n').slice(-10000).join('\n');
```
[the remark this answers is not in the document]
Comment (devversion) [reply dev-infra/utils/console.ts:185]: I guess I'm trying to figure out in which cases super old log output is helpful when the command conpleted successfully.

File: dev-infra/utils/console.ts [hunk 3]
```diff
@@ -155,3 +160,41 @@ function getLogLevel() {
...
+/** All text to write to the log file. */
+let LOGGED_TEXT = '';
+
+/** Enable the writing to the log file, setting the initial lines from the command execution. */
+export function enableFileLogging(argv: Arguments) {
+  /** The current date time when the file was initialized. */
+  const now = new Date();
+  /** Header line to separate command runs in log files. */
+  const headerLine = Array(100).fill('#').join('');
+  LOGGED_TEXT += `${headerLine}\nCommand: ${argv.$0} ${argv._.join(' ')}\nRan at: ${now}\n`;
```
[the remark this answers is not in the document]
Comment (josephperrott) [reply dev-infra/utils/console.ts:173]: Added in a check to only have the command run once.

File: dev-infra/utils/console.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (josephperrott) [reply dev-infra/utils/console.ts:185]: I'll switch over to just rewriting the whole file, and if we determine we would like to have it keep old outputs we can add that back in.

File: dev-infra/utils/console.ts [hunk 4]
```diff
@@ -155,3 +160,57 @@ function getLogLevel() {
...
+ * Enable writing the logged outputs to the log file on process exit, sets initial lines from the
+ * command execution, containing information about the timing and command parameters.
+ *
+ * This is expected to be called only once during a command run, and should be called by the
+ * middleware of yargs to enable the file logging before the rest of the command parsing and
+ * response is executed.
+ */
+export function enableFileLogging(argv: Arguments) {
+  if (FILE_LOGGING_ENABLED) {
+    debug('Skipping enabling of file logging as it is already enabled');
```
Comment (devversion) [inline dev-infra/utils/console.ts:184]: I think we should throw here as it signifies the following for which we should fix up the logic:

* We accidentally call this multiple times
* Potentially there is a need to capture multiple command outputs

**Review by devversion — APPROVED** [review]
LGTM. Personally I'd prefer the function to be named `captureLogOutputForCommand` but I don't feel too strong about it.

**Pull request closed without merging**